config/v1/types_cluster_version.go

```diff
@@ -0,0 +1,153 @@
+	//
+	// +optional
+	Channel string `json:"channel"`
+
+	// Overrides is list of overides for components that are managed by
+	// cluster version operator. Marking a component unmanaged will prevent
+	// the operator from creating or updating the object.
+	// +optional
+	Overrides []ComponentOverride `json:"overrides,omitempty"`
+}
+
+// ClusterVersionStatus reports the status of the cluster versioning,
+// including any upgrades that are in progress. The current field will
+// be set to whichever version the cluster is reconciling to, and the
+// conditions array will report whether the update succeeded, is in
+// progress, or is failing.
+// +k8s:deepcopy-gen=true
+type ClusterVersionStatus struct {
+	// current is the version that the cluster will be reconciled to. This
+	// value may be empty during cluster startup, and then will be set whenever
+	// a new update is being applied. Use the conditions array to know whether
+	// the update is complete.
+	Current Update `json:"current"`
+
+	// generation reports which version of the spec is being processed.
+	// If this value is not equal to metadata.generation, then the
+	// current and conditions fields have not yet been updated to reflect
+	// the latest request.
+	Generation int64 `json:"generation"`
+
+	// versionHash is a fingerprint of the content that the cluster will be
+	// updated with. It is used by the operator to avoid unnecessary work
+	// and is for internal use only.
+	VersionHash string `json:"versionHash"`
+
+	// conditions provides information about the cluster version. The condition
+	// "Available" is set to true if the desiredUpdate has been reached. The
+	// condition "Progressing" is set to true if an update is being applied.
+	// The condition "Failing" is set to true if an update is currently blocked
+	// by a temporary or permanent error. Conditions are only valid for the
+	// current desiredUpdate when metadata.generation is equal to
+	// status.generation.
+	Conditions []ClusterOperatorStatusCondition `json:"conditions"`
+
+	// AvailableUpdates contains the list of updates that are appropriate
+	// for this cluster. This list may be empty if no updates are recommended,
+	// if the update service is unavailable, or if an invalid channel has
+	// been specified.
+	AvailableUpdates []Update `json:"availableUpdates"`
+}
+
+// ClusterID is string RFC4122 uuid.
+type ClusterID string
+
+// ComponentOverride allows overriding cluster version operator's behavior
+// for a component.
+// +k8s:deepcopy-gen=true
+type ComponentOverride struct {
+	// Kind should match the TypeMeta.Kind for object.
+	Kind string `json:"kind"`
+
+	// The Namespace and Name for the component.
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+
+	// Unmanaged controls if cluster version operator should stop managing.
+	// Default: false
+	Unmanaged bool `json:"unmanaged"`
+}
+
+// URL is a thin wrapper around string that ensures the string is a valid URL.
+type URL string
```

Comment:
we haven't be assiduous about this. Would you like to be?   I have no strong feelings.

Reply:
Carry over from the CVO.  I think used judiciously it is good, but can be taken too far.